src/v/storage/segment_appender.cc

```diff
@@ -329,14 +334,33 @@ ss::future<> segment_appender::close() {
       .then([this] { return _out.close(); });
 }
 
-ss::future<> segment_appender::do_next_adaptive_fallocation() {
+ss::future<>
+segment_appender::do_next_adaptive_fallocation(size_t next_write_size) {
     auto step = _opts.resources.get_falloc_step(_opts.segment_size);
     if (step == 0) {
         // Don't fallocate.  This happens if we're low on disk, or if
         // the user has configured a 0 max falloc step.
         return ss::make_ready_future<>();
     }
 
+    // Optimize fallocation if we have a segment size hint.
+    if (_opts.segment_size) {
+        if (_fallocation_offset >= _opts.segment_size) {
+            // If we are already at the segment size, fallocate just to support
+            // the next write.
+            step = next_write_size;
+        } else if (_fallocation_offset + step >= _opts.segment_size) {
+            // If we are about to exceed the segment size, fallocate to the
+            // segment size and then some to support the extra write just before
+            // the segment is rolled.
+            step = *_opts.segment_size - _fallocation_offset
+                   + config::shard_local_cfg().kafka_batch_max_bytes();
```

Comment:
Can you explain this logic? Why did the previous version not work?

Reply:
You were right about 32 MiB max allocation step. So with a segment size of 128 MiB + 5% jitter it ended up doing 4 x 32 MiB fallocs, then an additional 32 MiB one (160 MiB).

This new logic captures this case too, thus capping the fallocation to the segment size hint plus 1 extra MiB (kafka_batch_max_bytes).

Reply:
I see, is `_opts.segment_size` the jittered version or the normal one?

I guess I am just wondering why we can't use `next_write_size` in either case? 

I am not sure `kafka_batch_max_bytes` is the right thing to do because some people set it super high to 10s or maybe even low hundreds of megabytes just to allow the occasional mega message.

Reply:
`_opts.segment_size` is the jittered version.

> I guess I am just wondering why we can't use next_write_size in either case?

Imagine the segment size is 130 MiB (128 MiB + 2 MiB jitter).
We did fallocate 4 times so we are at 128 MiB now (`_fallocation_offset`). How we should compute the last fallocate step?




```                     
                        + 130 MiB (segment size)
                        |
         + 128 MiB      |
 \\------|  ???         |
          
      a) |---|
             + next_write_size

      b) +---------------------------//-------------| 160 MiB
                                                    + default behavior (+32 MiB)

      c) +--------------|------| 131 MiB (segment size + kafka_batch_max_bytes)
                        +------+         (!) This PR.

```

Reply:
a) doesn't make sense

b) overallocates

c) this is what this PR does. it could just stop at 130 (segment size) and then fallocate once more for the next write but then you get one extra falloc + truncate

Reply:
I see, thanks for the diagram. Yeah it's tricky. 

Maybe another option would be `diff_to_max_segment + 2 * next_write_size`? I.e.: assuming all writes are similar sized so `2 * next_write_size` should be good enough to cover the final write.

Lastly, I think as a final step we should do `min(new_step, original_step)`. This would prevent edge cases with c) (like people having 100MB batch max bytes) and also cases where the next calculated step is small because of disk space etc.